### PR TITLE
Remote Joins: Create relationships across database and remote schemas

### PR DESCRIPTION
**Pull request by tirumaraiselvan**
### Try it out
The latest preview is available at https://hge-ci-pull-2392.herokuapp.com/

### Description

Remote Joins will enable joining data from your tables to remote schemas.

Remote joins will work similar to table relationships. You can define a relationship by giving:

- a name for the relationship
- a join configuration from table fields to remote schema fields.

Preview announcement: https://blog.hasura.io/remote-joins-a-graphql-api-to-join-database-and-other-data-sources/

### Affected components 

- Server
- Console
- Docs
- Build System
- Tests

### Related Issues

#1225
#1846
#2461 

### Limitations, known bugs & workarounds

- Transactions will not work for concurrent execution.
- Enum types are not supported for JOIN columns

### CODE REVIEW

**Comment by macalinao** [comment]
what's the ETA for this feature to get released?

**Comment by tirumaraiselvan** [comment]
> what's the ETA for this feature to get released?

Hi there! We are planning for end of this month.

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs [hunk 1]
```diff
@@ -142,19 +143,20 @@ runClearMetadata _ = do
 
 data ReplaceMetadata
   = ReplaceMetadata
-  { aqTables           :: ![TableMeta]
-  , aqFunctions        :: !(Maybe [QualifiedFunction])
-  , aqRemoteSchemas    :: !(Maybe [TRS.AddRemoteSchemaQuery])
-  , aqQueryCollections :: !(Maybe [DQC.CreateCollection])
-  , aqAllowlist        :: !(Maybe [DQC.CollectionReq])
+  { aqTables              :: ![TableMeta]
+  , aqFunctions           :: !(Maybe [QualifiedFunction])
```
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DDL/Metadata.hs]: I know this isn't new code, but anytime I see `Maybe [Something]`, I want to know (in a comment) the the semantic difference between `Just []` and `Nothing`. If there is no difference then it should probably just be a plain list.

File: server/src-lib/Data/Aeson/Ordered.hs [hunk 2]
```diff
@@ -0,0 +1,221 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+
+-- | A version of aeson that parses with key order preserved.
```
[the remark this answers is not in the document]
Comment (tirumaraiselvan) [reply server/src-lib/Data/Aeson/Ordered.hs]: Yes, this is basically a port of some aeson version which uses ordered map instead of hashmap which Chris had. I think we can pick out the version from some fpcomplete repo. This is used for maintaining the order of the query during serialization. 

I am not sure how much we should worry about this atm.

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs (hunk 1 above)
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/RQL/DDL/Metadata.hs]: I think most of these `Maybe [a]` is to maintain backward compatibility. We want to have backward compatible `replace_metadata` queries so the metadata from v1 (which didn't have `aqFunctions`) still needs to work with later versions.

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/RQL/DDL/Metadata.hs]: Is there a better way to solve this issue? Write our own parser instead of template haskell?

File: server/src-lib/Data/Aeson/Ordered.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tirumaraiselvan) [reply server/src-lib/Data/Aeson/Ordered.hs]: Although, if we can think of some other way of serialization then that might be more than better. Because ordered map is also not the best datastructure here as it is ordered "according to time of insert" rather than order according to some generic value. It works for us because the time property holds for us currently.

[the remark this answers is not in the document]
Comment (jberryman) [reply server/src-lib/Data/Aeson/Ordered.hs]: Thanks, that makes sense, and agree probably not a priority atm. Would be great to document this, esp your second comment there; I don't think it occurred to me that that's how it worked and that does sound fragile.

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs (hunk 1 above)
Comment (jberryman) [reply server/src-lib/Hasura/RQL/DDL/Metadata.hs]: That makes sense, thanks. Perhaps e.g. `Maybe (NonEmpty QualifiedFunction)` is what we want.

**Comment by KieranLafferty** [comment]
Any eta on merging this?

**Comment by niklabh** [comment]
throwing error 
TypeError: Cannot read property 'name' of undefined at https://graphql-engine-cdn.hasura.io/console/assets/versioned/pull2392-387a0053/main.js.gz:1:4972098 at Array.map ()

on metadata reload page

**Comment by Tbaut** [comment]
Super glad to see you work on this PR @rakeshkky 
We're happy to help for testing, we have a setup with 2 different remote schemas (pointing to a prisma and a self made graphql schemas).

Currently using `hasura/graphql-engine:pull2395-7ea7f82c` working well but super old.

Let us know when you think it's the right time to test :)

**Comment by rakeshkky** [comment]
@Tbaut You can tryout latest image commented by @hasura-bot to test the PR. Currently it is `hasura/graphql-engine:pull2392-4f48d2d2`. 

Please let us know your feedback in the comments. Thank you :slightly_smiling_face:

**Comment by Tbaut** [comment]
Great. 
I got the following error when launching `docker-compose up` with `hasura/graphql-engine:pull2392-1eba3caf`

```bash
graphql-engine_1  | {"internal":{"statement":"CREATE TABLE hdb_catalog.hdb_remote_relationship\n(\n  name TEXT NOT NULL,\n  table_schema name NOT NULL,\n  table_name name NOT NULL,\n  remote_schema TEXT NOT NULL,\n  configuration JSONB NOT NULL,\n  PRIMARY KEY (name, table_schema, table_name),\n  FOREIGN KEY (table_schema, table_name) REFERENCES hdb_catalog.hdb_table(table_schema, table_name) ON UPDATE CASCADE\n);\n","prepared":false,"error":{"exec_status":"FatalError","hint":null,"message":"relation \"hdb_remote_relationship\" already exists","status_code":"42P07","description":null},"arguments":[]},"path":"$","error":"postgres query error","code":"unexpected"}
graphql-engine_1  | {"type":"startup","timestamp":"2020-03-04T16:09:04.517+0000","level":"info","detail":{"kind":"server_configuration","info":{"live_query_options":{"batch_size":100,"refetch_delay":1},"transaction_isolation":"ISOLATION LEVEL READ COMMITTED","plan_cache_options":{"plan_cache_size":null},"enabled_log_types":["http-log","websocket-log","startup","webhook-log","query-log"],"server_host":"HostAny","enable_allowlist":false,"log_level":"info","auth_hook_mode":null,"use_prepared_statements":true,"unauth_role":"anonymous","stringify_numeric_types":false,"enabled_apis":["metadata","graphql","config","pgdump"],"enable_telemetry":true,"enable_console":true,"auth_hook":null,"jwt_secret":{"audience":null,"claims_format":null,"claims_namespace":null,"key":"<JWK REDACTED>","type":"RS256","issuer":null},"cors_config":{"allowed_origins":"*","disabled":false,"ws_read_cookie":null},"console_assets_dir":"/srv/console-assets","admin_secret_set":true,"port":8080}}}
graphql-engine_1  | {"type":"startup","timestamp":"2020-03-04T16:09:04.517+0000","level":"info","detail":{"kind":"postgres_connection","info":{"retries":1,"database_url":"postgres://postgres:...@postgres:5434/governance-discussion"}}}
postgres_1        | 2020-03-04 16:09:04.569 UTC [31] ERROR:  relation "hdb_remote_relationship" already exists
postgres_1        | 2020-03-04 16:09:04.569 UTC [31] STATEMENT:  CREATE TABLE hdb_catalog.hdb_remote_relationship
postgres_1        | 	(
postgres_1        | 	  name TEXT NOT NULL,
postgres_1        | 	  table_schema name NOT NULL,
postgres_1        | 	  table_name name NOT NULL,
postgres_1        | 	  remote_schema TEXT NOT NULL,
postgres_1        | 	  configuration JSONB NOT NULL,
postgres_1        | 	  PRIMARY KEY (name, table_schema, table_name),
postgres_1        | 	  FOREIGN KEY (table_schema, table_name) REFERENCES hdb_catalog.hdb_table(table_schema, table_name) ON UPDATE CASCADE
postgres_1        | 	);
```

I'll test with an empty DB but it'd be essential not to have to start from scratch.

**Comment by Tbaut** [comment]
Things worked great well with `hasura/graphql-engine:pull2392-1eba3caf` and starting with an empty DB.
I was able to do a migration, and play around without any hiccup :raised_hands: 

If the problem I raised earlier ( "hdb_remote_relationship" already exists) is a big pain to solve, we'll start with an empty DB then, don't want to bother you too much with that.

**Comment by rakeshkky** [comment]
@Tbaut The catalog migrate will be bad if you're running latest PR build on the database initiated older build of the same PR. The catalog version is subjected to change based on master branch.

**Comment by Tbaut** [comment]
@rakeshkky thanks for your answer. Do I understand correctly that until this PR is merged into master, we might have to start with an empty DB depending on where master goes?

If the answer to this question is yes, then I'd like to ask, when do you plan to merge this into master :) ?

**Comment by rakeshkky** [comment]
@Tbaut Yes. But there's a work around. 

1. Export metadata while on older version (important).
2. Drop catalog schemas using following SQL commands
```
DROP SCHEMA IF EXISTS hdb_views CASCADE;
DROP SCHEMA IF EXISTS hdb_catalog CASCADE;
```
3. Start the latest PR build
4. Import the metadata file exported in step 1

We're planning to merge this into master very soon.

**Comment by Tbaut** [comment]
That's 2 very good news :tada:, I'll try the workaround asap with the latest build and let you know how it went.

**Comment by Tbaut** [comment]
I just gave a shot at the workaround instructions you proposed.
I tried pasting the 2 `DROP SCHEMA` command in the SQL console from hasura, but both commands errored out
for the `hdb_views`: schema "hdb_views" does not exist
```js
{
    "statement": "CREATE VIEW \"hdb_views\".\"proposal_bot__insert__public__onchain_links\" AS SELECT * FROM \"public\".\"onchain_links\"",
    "prepared": false,
    "error": {
        "exec_status": "FatalError",
        "hint": null,
        "message": "schema \"hdb_views\" does not exist",
        "status_code": "3F000",
        "description": null
    },
    "arguments": []
}
```

for `hdb_catalog`: "relation "hdb_catalog.hdb_column" does not exist
```bash
unexpected : relation "hdb_catalog.hdb_column" does not exist
{
    "statement": "SELECT\n  t.table_schema,\n  t.table_name,\n  t.table_oid,\n  t.description,\n  coalesce(c.columns, '[]') as columns,\n  coalesce(f.constraints, '[]') as constraints,\n  coalesce(fk.fkeys, '[]') as foreign_keys\nFROM\n  (\n    SELECT\n      c.oid as table_oid,\n      c.relname as table_name,\n      n.nspname as table_schema,\n      pd.description as description\n    FROM\n      pg_catalog.pg_class c\n      JOIN pg_catalog.pg_namespace as n ON c.relnamespace = n.oid\n      LEFT JOIN pg_catalog.pg_description pd on (c.oid = pd.objoid and pd.objsubid = 0)\n  ) t\n  LEFT OUTER JOIN (\n    SELECT\n      table_schema,\n      table_name,\n      json_agg(\n        json_build_object(\n          'column_name', name,\n          'data_type', type,\n          'is_nullable', is_nullable :: boolean,\n          'ordinal_position', ordinal_position,\n          'references', primary_key_references,\n          'description', description\n        )\n      ) as columns\n    FROM\n      hdb_catalog.hdb_column\n    GROUP BY\n      table_schema,\n      table_name\n  ) c ON (\n    t.table_schema = c.table_schema\n    AND t.table_name = c.table_name\n  )\n  LEFT OUTER JOIN (\n    SELECT\n      tc.table_schema,\n      tc.table_name,\n      json_agg(\n        json_build_object(\n          'name',\n          tc.constraint_name,\n          'oid',\n          r.oid :: integer,\n          'type',\n          tc.constraint_type\n        )\n      ) as constraints\n    FROM\n      information_schema.table_constraints tc\n      JOIN pg_catalog.pg_constraint r ON tc.constraint_name = r.conname\n    GROUP BY\n      table_schema,\n      table_name\n  ) f ON (\n    t.table_schema = f.table_schema\n    AND t.table_name = f.table_name\n  )\n  LEFT OUTER JOIN (\n    SELECT\n      f.table_schema,\n      f.table_name,\n      json_agg(\n        json_build_object(\n          'table',\n          json_build_object(\n            'schema',\n            f.table_schema,\n            'name',\n            f.table_name\n          ),\n          'ref_table',\n          json_build_object(\n            'schema',\n            f.ref_table_table_schema,\n            'name',\n            f.ref_table\n          ),\n          'oid',\n          f.constraint_oid,\n          'constraint',\n          f.constraint_name,\n          'column_mapping',\n          f.column_mapping\n        )\n      ) as fkeys\n    FROM\n      hdb_catalog.hdb_foreign_key_constraint f\n    GROUP BY\n      table_schema,\n      table_name\n  ) fk ON (\n    fk.table_schema = t.table_schema\n    AND fk.table_name = t.table_name\n  )\nWHERE\n  t.table_schema NOT LIKE 'pg_%'\n  AND t.table_schema <> 'information_schema'\n  AND t.table_schema <> 'hdb_catalog'\n",
    "prepared": false,
    "error": {
        "exec_status": "FatalError",
        "hint": null,
        "message": "relation \"hdb_catalog.hdb_column\" does not exist",
        "status_code": "42P01",
        "description": null
    },
    "arguments": []
}
```

Thanks for your help in advance.

**Comment by rakeshkky** [comment]
@Tbaut I've updated the instructions https://github.com/hasura/graphql-engine/pull/2392#issuecomment-595180426 above. Ideally those commands should be executed in client other than Hasura.

**Comment by Tbaut** [comment]
Thanks @rakeshkky it works well using `psql` indeed. We'll continue active testing and report if there's anything unexpected :)

**Comment by Tbaut** [comment]
Hey @rakeshkky we just encountered the following with the latest docker image, with migrated data using the workaround (with `drop schema`) that you advised:

We added a webhook to hasura and needed to do a migration, but this did run well. We get:

```bash
FATA[0000] apply failed: [postgres-error] query execution failed ($.args[0].args)
File: '1572604593716_add replies/up.sql'
{
    "sql": "CREATE FUNCTION public.set_current_timestamp_updated_at() RETURNS trigger\n    LANGUAGE ... UPDATE RESTRICT ON DELETE RESTRICT;\n"
}
[42723] FatalError: function "set_current_timestamp_updated_at" already exists with same argument types 
```

This migration file `1572604593716` is the first and quite old that we have, this migration was applied, but it seems not to be detected.

Also the hasura docker console shows:
```bash
{"type":"metadata","timestamp":"2020-03-11T09:37:35.720+0000","level":"warn","detail":{"message":"Inconsistent Metadata!","info":{"objects":[{"definition":{"definition":{"timeout_seconds":60,"headers":[],"url_from_env":"HASURA_AUTH_SERVER_REMOTE_SCHEMA","forward_client_headers":true},"name":"auth","comment":null},"reason":"HTTP exception occurred while sending the request to http://auth-server-service:8010/auth/graphql","type":"remote_schema"},{"definition":{"remote_field":{"user":{"arguments":{"id":"$author_id"}}},"name":"author","hasura_fields":["author_id"],"remote_schema":"auth","table":{"schema":"public","name":"comments"}},"reason":"in table \"comments\": in remote relationship\"author\": cannot validate remote relationship because remote schema with name \"auth\" not found","type":"remote_relationship"},{"definition":{"remote_field":{"user":{"arguments":{"id":"$author_id"}}},"name":"author","hasura_fields":["author_id"],"remote_schema":"auth","table":{"schema":"public","name":"posts"}},"reason":"in table \"posts\": in remote relationship\"author\": cannot validate remote relationship because remote schema with name \"auth\" not found","type":"remote_relationship"}]}}}
graphql-engine_1  | {"type":"http-log","timestamp":"2020-03-11T09:37:35.720+0000","level":"info","detail":{"operation":{"query_execution_time":1.3899416e-2,"user_vars":{"x-hasura-role":"admin"},"request_id":"2901f9d9-d67a-45ca-9c8c-d81cbf434fdb","response_size":68,"request_read_time":5.387e-6},"http_info":{"status":200,"http_version":"HTTP/1.1","url":"/v1/query","ip":"127.0.0.1","method":"POST","content_encoding":"gzip"}}}
```

and the hasura console (on 8080) shows errors as well (these don't look helpful):
```bash
n/<@http://localhost:8080/console/assets/versioned/main.js.gz:1:4977470
n@http://localhost:8080/console/assets/versioned/main.js.gz:1:4977127
n@http://localhost:8080/console/assets/versioned/main.js.gz:1:4980880
n@http://localhost:8080/console/assets/versioned/main.js.gz:1:4983006
ho@http://localhost:8080/console/assets/versioned/vendor.js.gz:1:3046542
Xo@http://localhost:8080/console/assets/versioned/vendor.js.gz:1:3059013
fs@http://localhost:8080/console/assets/versioned/vendor.js.gz:1:3082377
ds@http://localhost:8080/console/assets/versioned/vendor.js.gz:1:3082765
Js@http://localhost:8080/console/assets/versioned/vendor.js.gz:1:3089943
```

Any idea what went wrong?

**Comment by Tbaut** [comment]
Just edited my message, I was mixing several things together sorry. I focused on the migration now.

**Comment by nathan-appere** [comment]
Just a note on the naming: I think talking about "JOIN" creates confusion about the feature, when we are actually talking about "preloading" / sub-queries.
Most of what makes JOIN powerful requires costly operations when targeting two remote schemas.
"Remote preload" would be more accurate.

**Comment by rakeshkky** [comment]
@jberryman Please take a look at the new changes of the PR and review. I made changes to merge the remote fields using internal annotated AST (`AnnFldG`) instead of GraphQL validation AST (`Field`) since we're soon going to get rid of Validation AST in parse-don't-validate refactor. The logic for extracting and resolving remote fields using internal annotated AST found in `Hasura.RQL.DML.RemoteJoin` module.

**Comment by Tbaut** [comment]
Hello @rakeshkky,

I've given a spin to the latest docker image and I get the following error in the UI
>DB schema loading failed
no such column exists : "name"

The docker shows this:
```bash
graphql-engine_1  | {"type":"http-log","timestamp":"2020-04-01T08:20:00.319+0000","level":"error","detail":{"operation":{"user_vars":{"x-hasura-role":"admin"},"error":{"path":"$.args[4].args.columns[0]","error":"no such column exists : \"name\"","code":"permission-denied"},"request_id":"ee623f9e-920c-434b-94c6-8ab8c4ca0e4b","response_size":106,"query":{"args":[{"args":{"cascade":false,"check_metadata_consistency":null,"sql":"select \n  COALESCE(\n    json_agg(\n      row_to_json(info)\n    ), \n    '[]' :: JSON\n  ) AS tables \nFROM \n  (\n    select \n      ist.table_schema, \n      ist.table_name,\n      ist.table_type,\n      obj_description(\n        (\n          quote_ident(ist.table_schema) || '.' || quote_ident(ist.table_name)\n        ):: regclass, \n        'pg_class'\n      ) AS comment, \n      COALESCE(json_agg(\n        DISTINCT row_to_json(is_columns) :: JSONB || jsonb_build_object(\n          'comment',\n          (\n            SELECT \n              pg_catalog.col_description(\n                c.oid, is_columns.ordinal_position :: int\n              ) \n            FROM \n              pg_catalog.pg_class c \n            WHERE \n              c.oid = (quote_ident(ist.table_schema) || '.' || quote_ident(ist.table_name)):: regclass :: oid\n              AND c.relname = is_columns.table_name\n          )\n        )\n      ) FILTER (WHERE is_columns.column_name IS NOT NULL), '[]' :: JSON) AS columns,\n      COALESCE(json_agg(\n        DISTINCT row_to_json(is_triggers) :: JSONB || jsonb_build_object(\n          'comment',\n          (\n            SELECT description FROM pg_description JOIN pg_trigger ON pg_description.objoid = pg_trigger.oid \n            WHERE \n              tgname = is_triggers.trigger_name \n              AND tgrelid = (quote_ident(is_triggers.event_object_schema) || '.' || quote_ident(is_triggers.event_object_table)):: regclass :: oid\n          )\n        )\n      ) FILTER (WHERE is_triggers.trigger_name IS NOT NULL), '[]' :: JSON) AS triggers,\n      row_to_json(is_views) AS view_info\n    FROM \n      information_schema.tables AS ist \n      LEFT OUTER JOIN information_schema.columns AS is_columns ON \n        is_columns.table_schema = ist.table_schema \n        AND is_columns.table_name = ist.table_name \n      LEFT OUTER JOIN information_schema.views AS is_views ON is_views.table_schema = ist.table_schema\n        AND is_views.table_name = ist.table_name\n      LEFT OUTER JOIN information_schema.triggers AS is_triggers ON \n        is_triggers.event_object_schema = ist.table_schema AND \n        is_triggers.event_object_table = ist.table_name\n    where (ist.table_schema='public') \n    GROUP BY \n      ist.table_schema, \n      ist.table_name,\n      ist.table_type,\n      is_views.*\n  ) AS info;","read_only":true},"version":1,"type":"run_sql"},{"args":{"where":null,"offset":null,"order_by":[{"type":"asc","column":"table_name"}],"columns":["table_schema","table_name","is_enum","configuration",{"where":null,"offset":null,"alias":null,"name":"primary_key","order_by":null,"columns":["*"],"limit":null},{"where":null,"offset":null,"alias":null,"name":"relationships","order_by":null,"columns":["*"],"limit":null},{"where":null,"offset":null,"alias":null,"name":"permissions","order_by":null,"columns":["*"],"limit":null},{"where":null,"offset":null,"alias":null,"name":"unique_constraints","order_by":null,"columns":["*"],"limit":null},{"where":null,"offset":null,"alias":null,"name":"check_constraints","order_by":[{"type":"asc","column":"constraint_name"}],"columns":["*"],"limit":null},{"where":null,"offset":null,"alias":null,"name":"computed_fields","order_by":[{"type":"asc","column":"computed_field_name"}],"columns":["*"],"limit":null}],"limit":null,"table":{"schema":"hdb_catalog","name":"hdb_table"}},"version":1,"type":"select"},{"args":{"cascade":false,"check_metadata_consistency":null,"sql":"select \n  COALESCE(\n    json_agg(\n      row_to_json(info)\n    ), \n    '[]' :: JSON\n  ) AS tables \nFROM \n  (\n    select\n      hdb_fkc.*, \n      fk_ref_table.table_name IS NOT NULL AS is_ref_table_tracked \n    from \n      hdb_catalog.hdb_table AS ist \n      JOIN hdb_catalog.hdb_foreign_key_constraint AS hdb_fkc ON hdb_fkc.table_schema = ist.table_schema \n      and hdb_fkc.table_name = ist.table_name \n      LEFT OUTER JOIN hdb_catalog.hdb_table AS fk_ref_table ON fk_ref_table.table_schema = hdb_fkc.ref_table_table_schema \n      and fk_ref_table.table_name = hdb_fkc.ref_table\n    where (ist.table_schema='public')\n  ) as info;","read_only":true},"version":1,"type":"run_sql"},{"args":{"cascade":false,"check_metadata_consistency":null,"sql":"select \n  COALESCE(\n    json_agg(\n      row_to_json(info)\n    ), \n    '[]' :: JSON\n  ) AS tables \nFROM \n  (\n    select DISTINCT ON (hdb_fkc.constraint_oid)\n      hdb_fkc.*, \n      fk_ref_table.table_name IS NOT NULL AS is_table_tracked,\n      hdb_uc.constraint_name IS NOT NULL AS is_unique\n    from \n      hdb_catalog.hdb_table AS ist \n      JOIN hdb_catalog.hdb_foreign_key_constraint AS hdb_fkc ON hdb_fkc.ref_table_table_schema = ist.table_schema \n      and hdb_fkc.ref_table = ist.table_name \n      LEFT OUTER JOIN hdb_catalog.hdb_table AS fk_ref_table ON fk_ref_table.table_schema = hdb_fkc.table_schema \n      and fk_ref_table.table_name = hdb_fkc.table_name\n      LEFT OUTER JOIN hdb_catalog.hdb_unique_constraint AS hdb_uc ON hdb_uc.table_schema = hdb_fkc.table_schema\n      and hdb_uc.table_name = hdb_fkc.table_name and ARRAY(select json_array_elements_text(hdb_uc.columns) ORDER BY json_array_elements_text) = ARRAY(select json_object_keys(hdb_fkc.column_mapping) ORDER BY json_object_keys)\n    where (ist.table_schema='public')\n  ) as info;","read_only":true},"version":1,"type":"run_sql"},{"args":{"where":{"_and":[{"table_schema":"public"}]},"offset":null,"order_by":[{"type":"asc","column":"name"}],"columns":["*.*","name"],"limit":null,"table":{"schema":"hdb_catalog","name":"hdb_remote_relationship"}},"version":1,"type":"select"}],"version":1,"type":"bulk"}},"http_info":{"status":400,"http_version":"HTTP/1.1","url":"/v1/query","ip":"127.0.0.1","method":"POST","content_encoding":null}}}
```
Can you help?

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+  , _rjSelSet        :: ![G.Field] -- ^ Selection set of remote field.
+  , _rjHasuraFields  :: !(HashSet FieldName) -- ^ Table fields.
+  , _rjFieldCall     :: !(NonEmpty FieldCall) -- ^ Remote fields.
+  , _rjRemoteSchema  :: !RemoteSchemaInfo -- ^ The remote schema server info.
+  , _rjPhantomFields :: ![PGColumnInfo]
+    -- ^ Hasura fields which are not in the selection set, but are required as
+    -- parameters to satisfy the remote join.
+  } deriving (Show, Eq)
+
+type RemoteJoins = NE.NonEmpty (FieldPath, NE.NonEmpty RemoteJoin)
```
[the remark this answers is not in the document]
Comment (rakeshkky) [reply server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:105]: @jberryman I've documented types in `Hasura.RQL.DML.RemoteJoin` module.

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs [hunk 4]
```diff
@@ -215,14 +219,19 @@ compositeValueToJSON = \case
...
+-- All such 'RemoteJoinField's of a pariticular remote schema are batched together
+-- and made GraphQL request to remote server to fetch remote join values.
 data RemoteJoinField
   = RemoteJoinField
-  { _rjfRemoteSchema :: !RemoteSchemaInfo
-  , _rjfAlias        :: !G.Alias
-  , _rjfField        :: !G.Field
-  , _rjfFieldCall    :: ![G.Name]
+  { _rjfRemoteSchema :: !RemoteSchemaInfo -- ^ The remote schema server info.
+  , _rjfAlias        :: !G.Alias -- ^ To level alias of the field
```
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:228]: "Top level"?  These comments are a huge improbement, thanks!

**Comment by Tbaut** [comment]
Can you help on [this problem](https://github.com/hasura/graphql-engine/pull/2392#issuecomment-607108013) @rakeshkky ?

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs (hunk 4 above)
Comment (rakeshkky) [reply server/src-lib/Hasura/RQL/DML/RemoteJoin.hs]: Yes, It is :smile: . I fixed the typo in latest commit. Now the PR is up to date with the master branch. Awaiting for your review. Thank you.

**Comment by rakeshkky** [comment]
Hi @Tbaut , I fixed the bug you reported in the latest commit.

**Comment by webdeb** [comment]
Guys, sorry for interrupting, but will there be a remote batch query option?
Let's say you are fetching some relationships remotely for N items? As I understood correctly hasura will call that remote service N times, or could we batch it to just a single remote query, by providing an `ids` parameter?

Btw. thanks for your work on this, can't wait to upgrade my server

**Comment by Tbaut** [comment]
Thanks @rakeshkky the lasr docker image works well :)

**Comment by Tbaut** [comment]
@rakeshkky I spoke too soon, the hasura docker is launching without any error, the hasura console showing and the data is still there.. however the remote fields aren't accessible with graphql queries!

GraphiQL doesn't show them although the remote relashonship are still visible in the UI, the remote schema url as well (reloading without error).

I can't see any error anywhere, but things aren't working. The last working image for us is `hasura/graphql-engine:pull2392-b7c91978`

**Comment by rakeshkky** [comment]
@webdeb The server actually batches the queries in a single request. We're trying hard to release this. Thank you.

**Comment by rakeshkky** [comment]
@Tbaut The [steps](https://github.com/hasura/graphql-engine/pull/2392#issuecomment-595180426) I mentioned to upgrade to latest PR build will no longer work in this PR since the logic of import/export metadata (for remote relationships) is changed. I'd suggest you to purge catalog data for remote relationships by performing `DELETE FROM hdb_catalog.hdb_remote_relationship` and re-define all remote relationships. 

Apologies for the inconvenience.

**Comment by weyert** [comment]
What's the best way to test this feature? I am seeing three PRs for this feature not sure which one to pick. Would it also be possible to combine it with query actions one?

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+-- We need to merge the input argument to where: {id : 1, name: "tiru"}
+mergeValue :: G.Value -> G.Value -> G.Value
+mergeValue lVal rVal = case (lVal, rVal) of
+  (G.VList (G.ListValueG l), G.VList (G.ListValueG r)) ->
+    G.VList $ G.ListValueG $ l <> r
+  (G.VObject (G.ObjectValueG l), G.VObject (G.ObjectValueG r)) ->
+    let fieldsToMap = Map.fromList . map (G._ofName &&& G._ofValue)
+    in G.VObject $ G.ObjectValueG $ map (uncurry G.ObjectFieldG) $ Map.toList $
+       Map.unionWith mergeValue (fieldsToMap l) (fieldsToMap r)
+  (l, _) -> l -- FIXME:- throw error for merging non-lists and non-objects
```
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:392]: I missed this the first time through, sorry. If we think this case is unreachable we can just make this `error "..."` and document why we expect the case is impossible (i.e. point to the validation we do etc.)

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+fieldsToRequest :: G.OperationType -> [G.Field] -> GQLReqParsed
```
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:394]: Since this is undocumented and used just once, can we move it into a where clause at the callsite in `fetchRemoteJoinFields`

File: server/src-lib/Hasura/RQL/DML/RemoteJoin.hs [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+     , _grQuery =
+         GQLExecDoc
+           [ G.ExecutableDefinitionOperation
+               (G.OperationDefinitionTyped
+                 ( emptyOperationDefinition
+                     {G._todSelectionSet = map G.SelectionField gfields}
+                 )
+               )
+           ]
+     , _grVariables = Nothing -- TODO: Put variables in here?
```
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:407]: What do you think about this `TODO`? I would need to look closer to have any insight

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/RQL/DML/RemoteJoin.hs:407]: I know what this is: if you have variables used in your remote joins field (from the client), then after validation phase it is expanded i.e no longer a variable.

But this is bad from an operation perspective because the remote server will see the expanded execution (so the variable values are out in the open). I tried to fix this in heterogenous requests PR here: https://github.com/hasura/graphql-engine/pull/3153/files#diff-ae73919a18ce8e018f747703c760d5c5R456

Basically, I un-expand the variables and plug it back.

**Comment by tavvy** [comment]
This is a great feature and have been testing using the commit build locally.
Is/would it be possible to specify the direction of the join? e.g

Hasura Main contains table `book_meta` which has a `book_id` field and a `price` field.

I add a graphql service via remote schema that has a `book` type with fields `book_id`, `title` etc

If I join these on a `book_id` field and query I would get something like

```
{
  book_meta {
    book_id
    price
    book {
      title
    }
  }
}
```
Ideally I would like to "reverse" the join as I may not hold a price for all books, but would still like to retrieve all books and optionally show the price e.g

```
{
  book {
    book_id
    title
    book_meta {
      price
    }
  }
}
```

Any thoughts? Perhaps I have mis-interpreted the feature in my local testing.

**Comment by ElixirMike** [comment]
Any timeline for a working version? Also, will we be able to use API to specify these remote relationships?

**Comment by esseswann** [comment]
I was playing with this pull request and I noticed these two problems:
First, seems that the nested `where` does not correctly work, and take notice that non of these are remote 
image
Expected behavior -- there should not be a question with team_answers.appeal = null

Second, the relationship tab in a view is crashing the frontend
image

**Comment by wawhal** [comment]
@esseswann it is a bug, will be fixed soon. Thanks for reporting.

**Comment by esseswann** [comment]
@wawhal Can confirm fix here https://github.com/hasura/graphql-engine/commit/63e528808c1f066e73f536a88a59feac2486b7d9, thanks

**Comment by esseswann** [comment]
@wawhal I think I found another problem, rather conceptual. I cannot use remote because remote relationship has arguments like ID! for example, which I cannot set for a column. Is there any way to solve this?

**Comment by Tbaut** [comment]
I've had the problem in the past @esseswann it's logged here: https://github.com/hasura/graphql-engine/issues/3695

TL;DR it's not possible at the moment, you need and Int!

**Comment by esseswann** [comment]
@Tbaut thanks, we'll watch that issue

File: console/src/components/Services/Data/TableRelationships/RemoteRelationships/components/RemoteRelEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+import styles from '../SchemaExplorer.scss';
+import {
+  RelName as RelNameTooltip,
+  RemoteSchema as RemoteSchemaTooltip,
+  Configuration as ConfigTooltip,
+} from '../Tooltips';
+import Explorer from './Explorer';
+
+type Props = {
+  table: any; // use "Table" type after ST is merged
```
Comment (beerose) [inline console/src/components/Services/Data/TableRelationships/RemoteRelationships/components/RemoteRelEditor.tsx:27]:
```suggestion
  table: any; // TODO: use "Table" type after ST is merged
```
So we can grep by "TODO"

**Comment by niklabh** [comment]
Congratulations on this merge

File: console/src/components/Services/Data/TableRelationships/RemoteRelationships/RemoteRelationships.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import styles from '../../TableModify/ModifyTable.scss';
+import { RemoteRelationshipServer } from './utils';
+import RemoteRelationshipList from './components/RemoteRelationshipList';
+import { fetchRemoteSchemas } from '../../../RemoteSchema/Actions';
+
+type Props = {
+  relationships: RemoteRelationshipServer[];
+  reduxDispatch: any;
```
[the remark this answers is not in the document]
Comment (wawhal) [reply console/src/components/Services/Data/TableRelationships/RemoteRelationships/RemoteRelationships.tsx:9]: Because there's another `dispatch` from `useReducer` floating around. This is just to differentiate the two.
